src/Storages/MergeTree/MergeTreeRangeReader.h

```diff
@@ -195,13 +209,10 @@ class MergeTreeRangeReader
 
         size_t countBytesInResultFilter(const IColumn::Filter & filter);
 
-        Columns columns;
-        size_t num_rows = 0;
-        bool need_filter = false;
+        bool need_filter = false; /// TODO: what exactly does it mean?
```

Comment:
As I remember, initially the idea was to reduce the number column filters.

For example, we have a query `select fat_string_column where y = 5 prewhere x = 42`.
Usually, we executed it like this:
```
--- prewhere part
read 'x'
calculate 'x = 42'
filter 'x' by 'x = 42'
read 'y', 'fat_string_column' (only for granules where 'x = 42')
filter 'y', 'fat_string_column' by 'x = 42'
--- where part
calculate 'y = 5'
filter 'fat_string_column' by 'x = 42'
```
Here we filter `fat_string_column` twice, and it may be not optimal.
There is an optimization: add a condition from PREWHERE to WHERE and skip filtering in PREWHERE.
Query will look like `select fat_string_column where y = 5 and x = 42 prewhere x = 42`.
The execution will look like:
```
--- prewhere part
read 'x'
calculate 'x = 42'
filter 'x' by 'x = 42'
read 'y', 'fat_string_column' (only for granules where 'x = 42')
// skip: filter 'y', 'fat_string_column' by 'x = 42'
--- where part
calculate 'y = 5 and x = 42'
filter 'fat_string_column' by 'y = 5' and 'x = 42'
```
Here we filter `fat_string_column` only once. Probably, we calculate `x = 42` twice here (or maybe not - it was possible to reuse calculated columns from prewhere some day), but in many cases it would be not so expensive as filtering.

Reply:
maybe @amosbird can verify if what I wrote is true

Reply:
@KochetovNicolai Perfect explanation. There is also one more benefit: delaying the filter reduces memory copy, because it's likely to filter the block again with additional filters in WHERE.  https://github.com/ClickHouse/ClickHouse/pull/7769